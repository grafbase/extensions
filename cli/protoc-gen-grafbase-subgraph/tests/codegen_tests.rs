#![allow(clippy::panic)]

use std::{collections::BTreeMap, fs, path::Path, process};

/// Parse a test file containing multiple proto files marked with //!file: comments
fn parse_test_file(content: &str) -> BTreeMap<String, String> {
    let mut files = BTreeMap::new();
    let mut current_file = None;
    let mut current_content = String::new();

    for line in content.lines() {
        if let Some(filename) = line.strip_prefix("//!file:") {
            // Save previous file if any
            if let Some(file) = current_file.take() {
                files.insert(file, current_content.trim().to_string());
                current_content.clear();
            }
            current_file = Some(filename.trim().to_string());
        } else if current_file.is_some() {
            current_content.push_str(line);
            current_content.push('\n');
        }
    }

    // Save last file
    if let Some(file) = current_file {
        files.insert(file, current_content.trim().to_string());
    }

    files
}

/// Write parsed proto files to a temporary directory
fn setup_proto_files(tmp_dir: &Path, files: &BTreeMap<String, String>) -> Vec<String> {
    let mut main_protos = Vec::new();

    for (filename, content) in files {
        let file_path = tmp_dir.join(filename);

        // Create parent directories if needed
        if let Some(parent) = file_path.parent() {
            fs::create_dir_all(parent).unwrap();
        }

        fs::write(&file_path, content).unwrap();

        // Track main proto files (not in subdirectories like grafbase/options.proto)
        if !filename.contains('/') || !filename.starts_with("grafbase/") {
            main_protos.push(filename.clone());
        }
    }

    main_protos
}

#[test]
fn codegen_tests() {
    let bin_path = env!("CARGO_BIN_EXE_protoc-gen-grafbase-subgraph");

    if cfg!(windows) {
        eprintln!("Skipping these tests on Windows, since there is no official protoc binary release.");
        return;
    }

    insta::glob!("testcases/*.proto", |test_path| {
        let snapshot_path = test_path.parent().unwrap();
        let test_name = test_path.file_stem().unwrap().to_str().unwrap();

        insta::with_settings!({
            snapshot_path => snapshot_path.to_str().unwrap(),
            prepend_module_to_snapshot => false,
            snapshot_suffix => "",
        }, {
            let test_content = fs::read_to_string(test_path).unwrap();
            let mut files = parse_test_file(&test_content);

            // If no //!file: markers, treat as single file for backwards compatibility
            if files.is_empty() {
                files.insert(
                    test_path.file_name().unwrap().to_str().unwrap().to_string(),
                    test_content,
                );
            }

            // Always insert the options.proto from the source of truth
            // This will override any inline options.proto in test files
            let options_proto_path = Path::new(env!("CARGO_MANIFEST_DIR"))
                .join("proto")
                .join("grafbase")
                .join("options.proto");
            let options_content = fs::read_to_string(options_proto_path).unwrap();
            files.insert("grafbase/options.proto".to_string(), options_content);

            let proto_tmp = tempfile::tempdir().unwrap();
            let main_protos = setup_proto_files(proto_tmp.path(), &files);

            let output_tmp = tempfile::tempdir().unwrap();

            for main_proto in main_protos {
                let mut cmd = process::Command::new("protoc");

                cmd.arg("--plugin")
                    .arg(bin_path)
                    .arg("--grafbase-subgraph_out")
                    .arg(output_tmp.path())
                    .arg("-I")
                    .arg(proto_tmp.path())
                    .arg(&main_proto)
                    .stderr(process::Stdio::inherit());

                let output = cmd.output().unwrap();

                assert!(
                    output.status.success(),
                    "Expected success for {}, got {}\n{}",
                    main_proto,
                    output.status,
                    std::str::from_utf8(&output.stderr).unwrap(),
                );
            }

            // Collect all generated .graphql files
            let mut graphql_files = BTreeMap::new();
            for entry in walkdir::WalkDir::new(output_tmp.path()) {
                let entry = entry.unwrap();

                if !entry.file_type().is_file() {
                    continue;
                }

                if let Some("graphql") = entry.path().extension().and_then(|s| s.to_str()) {
                    let filename = entry.file_name().to_str().unwrap().to_string();
                    let content = fs::read_to_string(entry.path()).unwrap();
                    graphql_files.insert(filename, content);
                }
            }

            // Debug: print what files were generated
            if test_name.contains("multi") {
                eprintln!("Generated files for {}: {:?}", test_name, graphql_files.keys().collect::<Vec<_>>());
            }

            // Concatenate all files in deterministic order
            let mut combined_output = String::new();
            for (filename, content) in graphql_files {
                if !combined_output.is_empty() {
                    combined_output.push_str("\n#--- ");
                    combined_output.push_str(&filename);
                    combined_output.push_str(" ---#\n\n");
                } else {
                    combined_output.push_str("#--- ");
                    combined_output.push_str(&filename);
                    combined_output.push_str(" ---#\n\n");
                }
                combined_output.push_str(&content);
            }

            insta::assert_snapshot!(test_name, combined_output);
        });
    });
}
