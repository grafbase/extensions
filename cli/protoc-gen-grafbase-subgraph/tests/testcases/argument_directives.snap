---
source: cli/protoc-gen-grafbase-subgraph/tests/codegen_tests.rs
expression: combined_output
input_file: cli/protoc-gen-grafbase-subgraph/tests/testcases/argument_directives.proto
---
#--- schema.graphql ---#

extend schema
  @link(url: "https://grafbase.com/extensions/grpc/0.2.0", import: ["@protoServices", "@protoEnums", "@protoMessages", "@grpcMethod"])
  @protoServices(
    definitions: [
      {
        name: "test.input_arg_directives.UserService"
        methods: [
          {
            name: "CreateUser"
            inputType: ".test.input_arg_directives.CreateUserRequest"
            outputType: ".test.input_arg_directives.User"
          }
          {
            name: "UpdateUser"
            inputType: ".test.input_arg_directives.UpdateUserRequest"
            outputType: ".test.input_arg_directives.User"
          }
          {
            name: "GetUser"
            inputType: ".test.input_arg_directives.GetUserRequest"
            outputType: ".test.input_arg_directives.User"
          }
          {
            name: "DeleteUser"
            inputType: ".test.input_arg_directives.DeleteUserRequest"
            outputType: ".test.input_arg_directives.DeleteUserResponse"
          }
        ]
      }
      {
        name: "test.input_arg_directives.AdminService"
        methods: [
          {
            name: "SearchUsers"
            inputType: ".test.input_arg_directives.SearchUsersRequest"
            outputType: ".test.input_arg_directives.SearchUsersResponse"
          }
        ]
      }
    ]
  )
  @protoMessages(
    definitions: [
      {
        name: ".test.input_arg_directives.User"
        fields: [
          {
            name: "id"
            number: 1
            repeated: false
            type: "string"
          }
          {
            name: "name"
            number: 2
            repeated: false
            type: "string"
          }
          {
            name: "email"
            number: 3
            repeated: false
            type: "string"
          }
        ]
      }
      {
        name: ".test.input_arg_directives.CreateUserRequest"
        fields: [
          {
            name: "name"
            number: 1
            repeated: false
            type: "string"
          }
          {
            name: "email"
            number: 2
            repeated: false
            type: "string"
          }
        ]
      }
      {
        name: ".test.input_arg_directives.UpdateUserRequest"
        fields: [
          {
            name: "id"
            number: 1
            repeated: false
            type: "string"
          }
          {
            name: "name"
            number: 2
            repeated: false
            type: "string"
          }
          {
            name: "email"
            number: 3
            repeated: false
            type: "string"
          }
        ]
      }
      {
        name: ".test.input_arg_directives.GetUserRequest"
        fields: [
          {
            name: "id"
            number: 1
            repeated: false
            type: "string"
          }
        ]
      }
      {
        name: ".test.input_arg_directives.DeleteUserRequest"
        fields: [
          {
            name: "id"
            number: 1
            repeated: false
            type: "string"
          }
        ]
      }
      {
        name: ".test.input_arg_directives.DeleteUserResponse"
        fields: [
          {
            name: "success"
            number: 1
            repeated: false
            type: "bool"
          }
        ]
      }
      {
        name: ".test.input_arg_directives.SearchUsersRequest"
        fields: [
          {
            name: "query"
            number: 1
            repeated: false
            type: "string"
          }
          {
            name: "limit"
            number: 2
            repeated: false
            type: "int32"
          }
        ]
      }
      {
        name: ".test.input_arg_directives.SearchUsersResponse"
        fields: [
          {
            name: "users"
            number: 1
            repeated: true
            type: ".test.input_arg_directives.User"
          }
          {
            name: "total_count"
            number: 2
            repeated: false
            type: "int32"
          }
        ]
      }
    ]
  )

type Query {
"""
Method with no input argument directives
"""
  test_input_arg_directives_UserService_GetUser(input: test_input_arg_directives_GetUserRequestInput): test_input_arg_directives_User @grpcMethod(service: "test.input_arg_directives.UserService", method: "GetUser")
"""
Query method with validation directive
"""
  test_input_arg_directives_AdminService_SearchUsers(input: test_input_arg_directives_SearchUsersRequestInput @constraint(maxLength: 100)): test_input_arg_directives_SearchUsersResponse @grpcMethod(service: "test.input_arg_directives.AdminService", method: "SearchUsers")
}

type Mutation {
"""
Method with a single directive on input argument
"""
  test_input_arg_directives_UserService_CreateUser(input: test_input_arg_directives_CreateUserRequestInput @constraint(minLength: 1)): test_input_arg_directives_User @grpcMethod(service: "test.input_arg_directives.UserService", method: "CreateUser")
"""
Method with multiple directives on input argument
"""
  test_input_arg_directives_UserService_UpdateUser(input: test_input_arg_directives_UpdateUserRequestInput @constraint(minLength: 1) @auth(rules: [{allow: owner}])): test_input_arg_directives_User @grpcMethod(service: "test.input_arg_directives.UserService", method: "UpdateUser")
"""
Method with both input argument directives and method directives
"""
  test_input_arg_directives_UserService_DeleteUser(input: test_input_arg_directives_DeleteUserRequestInput @constraint(format: "uuid")): test_input_arg_directives_DeleteUserResponse @grpcMethod(service: "test.input_arg_directives.UserService", method: "DeleteUser") @deprecated(reason: "Use removeUser instead")
}

"64 bit signed integer" scalar I64
"64 bit unsigned integer" scalar U64

input test_input_arg_directives_CreateUserRequestInput {
  name: String
  email: String
}

input test_input_arg_directives_UpdateUserRequestInput {
  id: String
  name: String
  email: String
}

input test_input_arg_directives_GetUserRequestInput {
  id: String
}

input test_input_arg_directives_DeleteUserRequestInput {
  id: String
}

input test_input_arg_directives_SearchUsersRequestInput {
  query: String
  limit: Int
}

"""
Messages
"""
type test_input_arg_directives_User {
  id: String
  name: String
  email: String
}

type test_input_arg_directives_DeleteUserResponse {
  success: Boolean
}

type test_input_arg_directives_SearchUsersResponse {
  users: [test_input_arg_directives_User!]
  total_count: Int
}
